### ISSUE TITLE
Could not unmount - unmount failed: exit status 32

### ISSUE THREAD

**Issue by martysweet** [kubernetes-sigs/aws-ebs-csi-driver#1663]
/kind bug

Related to [https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/1302](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/1302), we reproduced a similar scenario where the volume remains mounted to an instance and manual intervention is required to unblock it. Unlike the Global Mount code path in [https://github.com/kubernetes/kubernetes/pull/116138](https://github.com/kubernetes/kubernetes/pull/116138), which has been fixed in the latest patch versions, this issue took considerably more attempts to reproduce (around 2000 attach, 2000 detach operations).

**What happened?**

The Node CSI Driver failed to unmount the Pod Volume Mount/Bind, returning an RPC Error to the Kublet VolumeManager CSI handler. Like the other issue, this is probably occurring because it has already been unmounted.

```
ebs-plugin E0628 14:22:35.179292       1 driver.go:122] "GRPC error" err=<                                                                                                                                                                                                                                            
ebs-plugin     rpc error: code = Internal desc = Could not unmount "/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount": unmount failed: exit status 32
ebs-plugin     Unmounting arguments: /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount                                                                                                                                              
ebs-plugin     Output: umount: /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount: no mount point specified. 
```

 

Noticable bits of the Kubelet logs show the following: 

```
Jun 28 14:22:35 ip-10-227-165-112 kubelet-eks.daemon[8798]: E0628 14:22:35.184161    8798 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844 podName:59f3cba3-d6c1-40b8-a0df-71b83438c442 nodeName:}" failed. No retries permitted until 2023-06-28 14:22:35.684129616 +0000 UTC m=+12180.774870665 (durationBeforeRetry 500ms). Error: UnmountVolume.TearDown failed for volume "volume10" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844") pod "59f3cba3-d6c1-40b8-a0df-71b83438c442" (UID: "59f3cba3-d6c1-40b8-a0df-71b83438c442") : kubernetes.io/csi: Unmounter.TearDownAt failed: rpc error: code = Internal desc = Could not unmount "/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount": unmount failed: exit status 32
Jun 28 14:22:35 ip-10-227-165-112 kubelet-eks.daemon[8798]: Unmounting arguments: /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount
Jun 28 14:22:35 ip-10-227-165-112 kubelet-eks.daemon[8798]: Output: umount: /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount: no mount point specified.
Jun 28 14:22:35 ip-10-227-165-112 kubelet-eks.daemon[8798]: E0628 14:22:35.200894    8798 reconciler.go:208] "operationExecutor.UnmountVolume failed (controllerAttachDetachEnabled true) for volume \"volume10\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844\") pod \"59f3cba3-d6c1-40b8-a0df-71b83438c442\" (UID: \"59f3cba3-d6c1-40b8-a0df-71b83438c442\") : UnmountVolume.NewUnmounter failed for volume \"volume10\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844\") pod \"59f3cba3-d6c1-40b8-a0df-71b83438c442\" (UID: \"59f3cba3-d6c1-40b8-a0df-71b83438c442\") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json]: open /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json: no such file or directory" err="UnmountVolume.NewUnmounter failed for volume \"volume10\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844\") pod \"59f3cba3-d6c1-40b8-a0df-71b83438c442\" (UID: \"59f3cba3-d6c1-40b8-a0df-71b834
[kubelet.txt](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/files/11903571/kubelet.txt)
38c442\") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json]: open /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json: no such file or directory"
...
Jun 28 14:22:35 ip-10-227-165-112 kubelet-eks.daemon[8798]: E0628 14:22:35.505212    8798 reconciler.go:208] "operationExecutor.UnmountVolume failed (controllerAttachDetachEnabled true) for volume \"volume10\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844\") pod \"59f3cba3-d6c1-40b8-a0df-71b83438c442\" (UID: \"59f3cba3-d6c1-40b8-a0df-71b83438c442\") : UnmountVolume.NewUnmounter failed for volume \"volume10\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844\") pod \"59f3cba3-d6c1-40b8-a0df-71b83438c442\" (UID: \"59f3cba3-d6c1-40b8-a0df-71b83438c442\") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json]: open /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json: no such file or directory" err="UnmountVolume.NewUnmounter failed for volume \"volume10\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844\") pod \"59f3cba3-d6c1-40b8-a0df-71b83438c442\" (UID: \"59f3cba3-d6c1-40b8-a0df-71b83438c442\") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json]: open /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json: no such file or directory"
Jun 28 14:22:35 ip-10-227-165-112 kubelet-eks.daemon[8798]: E0628 14:22:35.606029    8798 reconciler.go:208] "operationExecutor.UnmountVolume failed (controllerAttachDetachEnabled true) for volume \"volume10\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844\") pod \"59f3cba3-d6c1-40b8-a0df-71b83438c442\" (UID: \"59f3cba3-d6c1-40b8-a0df-71b83438c442\") : UnmountVolume.NewUnmounter failed for volume \"volume10\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844\") pod \"59f3cba3-d6c1-40b8-a0df-71b83438c442\" (UID: \"59f3cba3-d6c1-40b8-a0df-71b83438c442\") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json]: open /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json: no such file or directory" err="UnmountVolume.NewUnmounter failed for volume \"volume10\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-01e3c874d68c54844\") pod \"59f3cba3-d6c1-40b8-a0df-71b83438c442\" (UID: \"59f3cba3-d6c1-40b8-a0df-71b83438c442\") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json]: open /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/vol_data.json: no such file or directory"

```

Attached is a more complete log around this time. >> https://github.com/kubernetes-sigs/aws-ebs-csi-driver/files/11903580/kubelet.txt

**What you expected to happen?**

If the volume has been previously unmounted by a job, the driver should succeed with the unmount request. 

There are currently checks for if `not mounted` is present in the response.
`no mount point specified` is not handled.

[https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/release-1.19/pkg/driver/mount_linux.go#L144-L156C2](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/release-1.19/pkg/driver/mount_linux.go#L144-L156C2)

**How to reproduce it (as minimally and precisely as possible)?**

Move a single StatefulSet (with volumes) between two nodes (Ubuntu 20.04.6 AMI).
Using [https://github.com/martysweet/eks-ebs-node-detachment-reproduction](https://github.com/martysweet/eks-ebs-node-detachment-reproduction) to automate this process continuously, and running for ~4 hours. In this example, we used 20 volumes and ~100 iterations.

**Environment**

```bash
root@ip-10-227-165-112:~# uname -a
Linux ip-10-227-165-112 5.15.0-1038-aws #43~20.04.1-Ubuntu SMP Fri Jun 2 17:10:57 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

root@ip-10-227-165-112:~# cat /etc/issue
Ubuntu 20.04.6 LTS \n \l

root@ip-10-227-165-112:~# kubelet --version
Kubernetes v1.25.10
```

- Driver version: `1.19.0`
- Node AMI: `ami-04dad98d97f61affd` `ubuntu-eks/k8s_1.25/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20230623`

**Anything else we need to know?**

- As mentioned in [https://github.com/kubernetes/kubernetes/pull/116138#issuecomment-1451129765](https://github.com/kubernetes/kubernetes/pull/116138#issuecomment-1451129765), there possibly exists a TOCTOU bug in the CSI job scheduler where it’s possible for an unmount call to be scheduled twice. In this instance, the unmount was for the pod/bind path (not the global mount)
- The `TearDownAt` function calls the CSI Driver [https://github.com/kubernetes/kubernetes/blob/7c41bbbbdf41da1f366db4237cf25dbe0f05dd3a/pkg/volume/csi/csi_mounter.go#L417-L419](https://github.com/kubernetes/kubernetes/blob/7c41bbbbdf41da1f366db4237cf25dbe0f05dd3a/pkg/volume/csi/csi_mounter.go#L417-L419)
- Invoking [https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/release-1.19/pkg/driver/mount_linux.go#L144-L156C2](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/release-1.19/pkg/driver/mount_linux.go#L144-L156C2)
- The CSI Driver runs umount using the `mount-utils` package [https://github.com/kubernetes/mount-utils/blob/master/mount_linux.go#L359-L367](https://github.com/kubernetes/mount-utils/blob/master/mount_linux.go#L359-L367) and results in a return code of 32
- The error is bubbled up and results in the TearDown failing.
- The job is retried and the kubelet get stuck in the following code, as the pod mount is unmounted and the vol_data.json is not present at the pod mount location
[https://github.com/kubernetes/kubernetes/blob/8f09151d9c11b0219450ca8160fb770512fe68e5/pkg/volume/csi/csi_plugin.go#L454-L458](https://github.com/kubernetes/kubernetes/blob/8f09151d9c11b0219450ca8160fb770512fe68e5/pkg/volume/csi/csi_plugin.go#L454-L458)

There are possibly a few issues here:

- Jobs being scheduled twice by the Kublet VolumeManager and then running twice. However, in the event of an RPC timeout, as mentioned [https://github.com/kubernetes/kubernetes/pull/116138#issuecomment-1451129765](https://github.com/kubernetes/kubernetes/pull/116138#issuecomment-1451129765), this might happen regardless.
- The existing logic in the `ebs-csi-driver` checking for `not mounted` does not capture all edgecases
- Maybe the `csi_plugin.go` should fail-fast in the event that vol_data.json is missing. This is how the global mount is handled [https://github.com/kubernetes/kubernetes/blob/56f50bd0aecaa5877cf7b37bae6d5884007210b3/pkg/volume/csi/csi_attacher.go#L598-L601](https://github.com/kubernetes/kubernetes/blob/56f50bd0aecaa5877cf7b37bae6d5884007210b3/pkg/volume/csi/csi_attacher.go#L598-L601)

We don’t have the internal ins-and-outs knowledge of the CSI driver contracts, so we’ve decided to raise the issue here for initial discussion and visibility - as users would see this error first in the ebs node logs.

**********************************Further Debugging**********************************

Running umount manually had some interesting results. The folder structure has been left in a state where the pod folder exists, but the pvc folder has been removed (presumably from the job that successfully unmounted the pod volume).

```sql
root@ip-10-227-165-112:~# mkdir new
root@ip-10-227-165-112:~# umount new
umount: new: not mounted.

root@ip-10-227-165-112:~# echo $?
32

root@ip-10-227-165-112:~# ls -lh /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/
total 0

root@ip-10-227-165-112:~# umount /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609
umount: /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609: no mount point specified.

root@ip-10-227-165-112:~# umount /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount
umount: /var/lib/kubelet/pods/59f3cba3-d6c1-40b8-a0df-71b83438c442/volumes/kubernetes.io~csi/pvc-b097ef05-7813-4813-b245-23af831d9609/mount: no mount point specified.

root@ip-10-227-165-112:~# echo $?
32

root@ip-10-227-165-112:~# umount -V
umount from util-linux 2.34 (libmount 2.34.0: selinux, smack, btrfs, namespaces, assert, debug)

```

**Comment by martysweet** [comment]
[kubelet.txt](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/files/11903580/kubelet.txt)

**Comment by DLakin01** [comment]
+1

**Comment by jbma2z** [comment]
Possibly related to https://github.com/kubernetes/kubernetes/pull/116138

**Comment by martysweet** [comment]
> Unlike the Global Mount code path in https://github.com/kubernetes/kubernetes/pull/116138, which has been fixed in the latest patch versions, this issue took considerably more attempts to reproduce (around 2000 attach, 2000 detach operations).

@jbma2z Unfortunately, this was tested with the patch from that PR

**Comment by jbma2z** [comment]
@martysweet I did an extensive test run and was not able to repro this on EKS 1.27 with EBS CSI driver v1.21.0-eksbuild.1 after ~17000 iterations (over 2 weeks).

**Comment by jbma2z** [comment]
@martysweet This was on latest EKS optimized AL2 nodes at the time of testing, eg: `amazon-eks-node-1.27-v20230816`.

I did see a repro of the issue as described on EKS 1.24 with an old version, 1.11.2, of the aws-ebs-csi-driver using the EKS optimized AL2 AMI. 

I was not able to repro the issue on either cluster using the 1.21.0-eksbuild.1 version of the driver with EKS optimized AL2 images.

**Comment by jbma2z** [comment]
I only tested with the EKS Optimized Amazon Linux 2 images. Any issues with the Canonical images would need to be brought up with them, though I think the team responsible for the drivers might be able to address them as well. I can do another run on Ubuntu based images to see if there are any differences in the results. I will report back here when I finish those.

**Comment by jbma2z** [comment]
Using EBS CSI Driver v1.23.0-eksbuild.1 and ubuntu-eks/k8s_1.24/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20230922, I completed over 5k iterations without issue.

**Comment by martysweet** [comment]
Hi @jbma2z,

I have tested this again with the following:

> 2x c6a.large
> kubelet v1.24.17
> aws-ebs-csi-driver 1.23.0
> ubuntu-eks/k8s_1.24/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20231006

I could not reproduce this issue over 308 pod migrations attached to 20 volumes.
~6160 mounts and ~6160 dismounts.

Most runs complete within 95 seconds from Drain -> Running.
A single iteration took 335 seconds, but all the volumes were mounted successfully (within 105 secs, so I don't think this is the same as the unmount issue we saw in this ticket.

I'm happy for this to be closed - thanks for looking into it!

Marty
